Type hints only support bare `@environ.config` without () or passing args

With (probably) https://github.com/hynek/environ-config/commit/683b1294af08bfce495f8bb7ca3723ea34c49894 typehints for the public apis where introduced (which i really like).

However it seems like that these typehints produce issues if you use `environ.config()`.

Using the following example (which works technically fine) we see issues when validating it with mypy.

```python
import environ

@environ.config()
class TestConfig:
    test_var = environ.var()

print(environ.to_config(TestConfig))
```

mypy outputs: 

```
$ mypy test.py 
test.py:4: error: <nothing> not callable  [misc]
Found 1 error in 1 file (checked 1 source file)
```


This can also be reproduced allthought with a slightly different error:

```python
import environ

class TestConfig:
    test_var = environ.var()

Test2 = environ.config(TestConfig, prefix="a")

print(environ.to_config(Test2))
```

```
$ mypy test.py
test.py:8: error: Argument 1 to "to_config" has incompatible type "TestConfig"; expected "Type[<nothing>]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,12 @@ Whenever there is a need to break compatibility, it is announced here in the cha
 
 ## [Unreleased](https://github.com/hynek/environ-config/compare/23.1.0...HEAD)
 
+### Fixed
+
+- Type hints for `environ.config()` now allow for arguments (e.g.
+  `@environ.config(prefix="")`.
+
+
 ## [23.1.0](https://github.com/hynek/environ-config/compare/22.1.0...23.1.0) - 2023-01-27
 
 ### Backwards-incompatible Changes

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,8 +20,8 @@ Whenever there is a need to break compatibility, it is announced here in the cha
 
 ### Fixed
 
-- Type hints for `environ.config()` now allow for arguments (e.g.
-  `@environ.config(prefix="")`.
+- Type hints for `environ.config()` now allow for arguments (e.g. `@environ.config(prefix="")`).
+  [#56](https://github.com/hynek/environ-config/issues/56)
 
 
 ## [23.1.0](https://github.com/hynek/environ-config/compare/22.1.0...23.1.0) - 2023-01-27

--- a/src/environ/_environ_config.py
+++ b/src/environ/_environ_config.py
@@ -17,7 +17,7 @@ from __future__ import annotations
 import logging
 import os
 
-from typing import Any, Callable, TypeVar
+from typing import Any, Callable, TypeVar, overload
 
 import attr
 
@@ -60,13 +60,30 @@ RAISE = Raise()
 T = TypeVar("T")
 
 
+@overload
 def config(
-    maybe_cls: type[T] | None = None,
+    *,
     prefix: str | Sentinel = PREFIX_NOT_SET,
     from_environ: str = "from_environ",
     generate_help: str = "generate_help",
     frozen: bool = False,
-) -> T:
+) -> Callable[[T], T]:
+    ...
+
+
+@overload
+def config(maybe_cls: type[T]) -> T:
+    ...
+
+
+def config(
+    maybe_cls: type[T] | None = None,
+    *,
+    prefix: str | Sentinel = PREFIX_NOT_SET,
+    from_environ: str = "from_environ",
+    generate_help: str = "generate_help",
+    frozen: bool = False,
+) -> T | Callable[[T], T]:
     """
     Make a class a configuration class.
 

--- a/src/environ/_environ_config.py
+++ b/src/environ/_environ_config.py
@@ -67,12 +67,12 @@ def config(
     from_environ: str = "from_environ",
     generate_help: str = "generate_help",
     frozen: bool = False,
-) -> Callable[[T], T]:
+) -> Callable[[type[T]], type[T]]:
     ...
 
 
 @overload
-def config(maybe_cls: type[T]) -> T:
+def config(maybe_cls: type[T]) -> type[T]:
     ...
 
 
@@ -83,7 +83,7 @@ def config(
     from_environ: str = "from_environ",
     generate_help: str = "generate_help",
     frozen: bool = False,
-) -> T | Callable[[T], T]:
+) -> type[T] | Callable[[type[T]], type[T]]:
     """
     Make a class a configuration class.
 

--- a/typing_examples.py
+++ b/typing_examples.py
@@ -58,3 +58,13 @@ s: str = cfg.secret
 ds: str = cfg.d_secret
 vs: str = cfg.v_secret
 as_: str = cfg.a_secret
+
+
+@environ.config(prefix="")
+class ConfigNoPrefix:
+    test_var = environ.var()
+
+
+@environ.config()
+class ConfigEmptyParens:
+    test_var = environ.var()

